kernel/comps/keyboard/src/i8042_chip/mouse.rs

```diff
@@ -0,0 +1,479 @@
+
+/// Mouse packet state machine.
+static MOUSE_PACKET_STATE: SpinLock<PacketState, LocalIrqDisabled> =
+    SpinLock::new(PacketState::new());
+
+trait MouseCommand {
+    const CMD_BYTE: u8;
+    const DATA_LEN: usize;
+    const RES_LEN: usize;
+}
+
+struct InitCtx<'a>(&'a mut I8042Controller);
+
+impl InitCtx<'_> {
+    fn get_device_id(&mut self) -> Result<u8, I8042ControllerError> {
+        let mut buf = [0u8; GetDeviceId::RES_LEN];
+        self.command::<GetDeviceId>(&[], &mut buf)?;
+        Ok(buf[0])
+    }
+
+    fn enable_intellimouse(&mut self) -> Result<(), I8042ControllerError> {
+        const SAMPLE_RATE_200: u8 = 200;
+        const SAMPLE_RATE_100: u8 = 100;
+        const SAMPLE_RATE_80: u8 = 80;
+
+        // Set sample rate with 200, then 100, then 80.
+        self.set_sample_rate(SAMPLE_RATE_200)?;
+        self.set_sample_rate(SAMPLE_RATE_100)?;
+        self.set_sample_rate(SAMPLE_RATE_80)?;
+        Ok(())
+    }
+
+    fn set_sample_rate(&mut self, rate: u8) -> Result<(), I8042ControllerError> {
+        let mut buf = [0u8; SetSampleRate::RES_LEN];
+        self.command::<SetSampleRate>(&[rate], &mut buf)?;
+        Ok(())
+    }
+
+    fn enable_data_reporting(&mut self) -> Result<(), I8042ControllerError> {
+        let mut buf = [0u8; EnableDataReporting::RES_LEN];
+        self.command::<EnableDataReporting>(&[], &mut buf)?;
+        Ok(())
+    }
+
+    fn command<C: MouseCommand>(
+        &mut self,
+        args: &[u8],
+        out: &mut [u8],
+    ) -> Result<(), I8042ControllerError> {
+        debug_assert_eq!(args.len(), C::DATA_LEN);
+        debug_assert_eq!(out.len(), C::RES_LEN);
+
+        self.0.write_to_second_port(C::CMD_BYTE)?;
+        if self.0.wait_and_recv_data()? != PS2_ACK {
+            return Err(I8042ControllerError::DeviceResetFailed);
+        }
+
+        for &arg in args.iter().take(C::DATA_LEN) {
+            self.0.write_to_second_port(arg)?;
+            if self.0.wait_and_recv_data()? != PS2_ACK {
+                return Err(I8042ControllerError::DeviceResetFailed);
+            }
+        }
+
+        for slot in out.iter_mut().take(C::RES_LEN) {
+            *slot = self.0.wait_and_recv_data()?;
+        }
+
+        Ok(())
+    }
+}
+
+mod cmd {
+    use super::MouseCommand;
+
+    macro_rules! define_commands {
+        (
+            $(
+                $name:ident, $cmd:literal, fn([u8; $dlen:literal]) -> [u8; $rlen:literal];
+            )*
+        ) => {
+            $(
+                pub(super) struct $name;
+                impl MouseCommand for $name {
+                    const CMD_BYTE: u8 = $cmd;
+                    const DATA_LEN: usize = $dlen;
+                    const RES_LEN: usize = $rlen;
+                }
+            )*
+        };
+    }
+
+    define_commands! {
+        GetDeviceId, 0xF2, fn([u8; 0]) -> [u8; 1];
+        SetSampleRate, 0xF3, fn([u8; 1]) -> [u8; 0];
+        EnableDataReporting, 0xF4, fn([u8; 0]) -> [u8; 0];
+    }
+}
+use cmd::*;
```

Comment:
https://github.com/asterinas/asterinas/pull/2479#discussion_r2548905816
> Using `cmd::GetDeviceId` is more clearer than a global import?

I mean we can use `cmd::SetSampleRate` and remove this wildcard import.